src/subscription_manager/scripts/rhsmcertd_worker.py

```diff
@@ -89,82 +99,168 @@ def _collect_cloud_info(cloud_list: list, log) -> dict:
-    # because we want to have more sensitive detection in this case
-    # (automatic registration is more important than reporting of facts)
+    # Try to detect cloud provider first. We use lower threshold in this case, not the
+    # default, because we want to have more sensitive detection in this case;
+    # automatic registration is more important than reporting of facts.
     cloud_list = detect_cloud_provider(threshold=0.3)
     if len(cloud_list) == 0:
-        log.warning("This system does not run on any supported cloud provider. Skipping auto-registration")
-        sys.exit(-1)
+        log.warning(
+            "This system does not run on any supported cloud provider. "
+            "Automatic registration cannot be performed."
+        )
+        sys.exit(1)
 
-    # When some cloud provider(s) was detected, then try to collect metadata
-    # and signature
-    cloud_info = _collect_cloud_info(cloud_list, log)
+    # When some cloud provider(s) were detected, then try to collect metadata and signature
+    cloud_info = _collect_cloud_info(cloud_list)
     if len(cloud_info) == 0:
-        log.warning("It was not possible to collect any cloud metadata. Unable to perform auto-registration")
-        sys.exit(-1)
+        log.warning("Cloud metadata could not be collected. Unable to perform automatic registration.")
+        sys.exit(1)
 
     # Get connection not using any authentication
-    cp = cp_provider.get_no_auth_cp()
+    uep: "UEPConnection" = cp_provider.get_no_auth_cp()
 
-    # Try to get JWT token from candlepin (cloud registration adapter)
+    # Obtain automatic registration token
     try:
-        jwt_token = cp.getJWToken(
+        token: Dict[str, str] = cache.CloudTokenCache.get(
+            uep=uep,
             cloud_id=cloud_info["cloud_id"],
             metadata=cloud_info["metadata"],
             signature=cloud_info["signature"],
         )
-    except Exception as err:
-        log.error("Unable to get JWT token: {err}".format(err=str(err)))
-        log.warning("Canceling auto-registration")
-        sys.exit(-1)
+    except Exception:
+        log.exception(
+            "Anonymous cloud token could not be obtained. Unable to perform automatic registration."
+        )
+        sys.exit(1)
+
+    if token["tokenType"] == "CP-Cloud-Registration":
+        try:
+            _auto_register_standard(uep=uep, token=token)
+        except Exception:
+            log.exception("Standard automatic registration failed.")
+            sys.exit(1)
+        else:
+            log.info("Standard automatic registration was successful.")
+            sys.exit(0)
+
+    if token["tokenType"] == "CP-Anonymous-Cloud-Registration":
+        try:
+            _auto_register_anonymous(uep=uep, token=token)
+            cache.CloudTokenCache.delete_cache()
+        except Exception:
+            log.exception("Anonymous automatic registration failed.")
+            sys.exit(1)
```

Comment:
Maybe we could return here some other error value than `1` to be able distinguish different error states in `rhsmcertd.c`.

Reply:
I've pushed an additional commit. Let me know what you think.

Reply:
Very well! :+1: